File: components/tc-api/app/services/comment_service.py
from typing import List, Tuple, Optional
from datetime import datetime
from bson import ObjectId
from pymongo import DESCENDING

from app.core.database import get_database
from app.models.comment import CommentCreate, CommentInDB

class CommentService:
    def __init__(self):
        self.db = get_database()

    async def get_solution_comments(
        self,
        solution_slug: str,
        skip: int = 0,
        limit: int = 20,
        sort_by: str = "created_at"
    ) -> Tuple[List[CommentInDB], int]:
        """Get all comments for a solution with pagination.
        Comments are sorted by created_at in descending order (newest first)."""
        query = {"solution_slug": solution_slug}
        sort_field = sort_by
        cursor = self.db.comments.find(query).sort(sort_field, DESCENDING).skip(skip).limit(limit)
        comments = [CommentInDB(**comment) async for comment in cursor]
        total = await self.db.comments.count_documents(query)
        return comments, total

    async def get_comment_by_id(self, comment_id: str) -> Optional[CommentInDB]:
        """Get a specific comment by ID"""
        comment = await self.db.comments.find_one({"id": comment_id})
        if comment:
            return CommentInDB(**comment)
        return None

    async def create_comment(
        self,
        solution_slug: str,
        comment: CommentCreate,
        username: str
    ) -> CommentInDB:
        """Create a new comment"""
        now = datetime.utcnow().isoformat()
        comment_dict = comment.model_dump()
        new_comment = {
            "id": str(ObjectId()),
            "solution_slug": solution_slug,
            "username": username,
            "content": comment_dict["content"],
            "created_at": now,
            "updated_at": now
        }
        await self.db.comments.insert_one(new_comment)
        return CommentInDB(**new_comment)

    async def update_comment(
        self,
        comment_id: str,
        content: str,
        username: str
    ) -> Optional[CommentInDB]:
        """Update a comment's content"""
        now = datetime.utcnow().isoformat()
        result = await self.db.comments.update_one(
            {"id": comment_id, "username": username},  # Only allow update by comment author
            {
                "$set": {
                    "content": content,
                    "updated_at": now
                }
            }
        )
        if result.modified_count == 0:
            return None
        return await self.get_comment_by_id(comment_id)

    async def delete_comment(self, comment_id: str, username: str) -> bool:
        """Delete a comment"""
        result = await self.db.comments.delete_one({
            "id": comment_id,
            "username": username  # Only allow deletion by comment author
        })
        return result.deleted_count > 0 

File: components/compass-ui/src/app/core/services/comment.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { environment } from "../../../environments/environment";

export interface Comment {
  _id: string;
  content: string;
  username: string;
  full_name: string;
  created_at: string;
  solution_slug: string;
  type: "OFFICIAL" | "USER";
  is_adopted_user?: boolean;
  created_by?: string;
  updated_at?: string;
  updated_by?: string;
}

export interface CommentResponse {
  success: boolean;
  data: Comment[];
  total: number;
}

@Injectable({
  providedIn: "root",
})
export class CommentService {
  constructor(private http: HttpClient) {}

  // Get all comments (admin only)
  getAllComments(
    skip: number = 0,
    limit: number = 10,
    filters?: {
      solution_slug?: string;
      type?: "OFFICIAL" | "USER";
    }
  ): Observable<CommentResponse> {
    let url = `${environment.apiUrl}/comments/?skip=${skip}&limit=${limit}`;
    
    if (filters) {
      if (filters.solution_slug) {
        url += `&solution_slug=${filters.solution_slug}`;
      }
      if (filters.type) {
        url += `&type=${filters.type}`;
      }
    }
    
    return this.http.get<CommentResponse>(url);
  }

  // Get comments for a specific solution
  getSolutionComments(
    slug: string,
    skip: number = 0,
    limit: number = 10,
    type?: "OFFICIAL" | "USER"
  ): Observable<CommentResponse> {
    let url = `${environment.apiUrl}/comments/solution/${slug}?skip=${skip}&limit=${limit}`;
    if (type) {
      url += `&type=${type}`;
    }
    return this.http.get<CommentResponse>(url);
  }

  // Get my comments
  getMyComments(
    skip: number = 0,
    limit: number = 10
  ): Observable<CommentResponse> {
    return this.http.get<CommentResponse>(
      `${environment.apiUrl}/comments/my/?skip=${skip}&limit=${limit}`
    );
  }

  // Add a comment to a solution
  addComment(
    slug: string, 
    data: { content: string; is_adopted_user: boolean }
  ): Observable<any> {
    return this.http.post<any>(
      `${environment.apiUrl}/comments/solution/${slug}`,
      data
    );
  }

  // Update a comment
  updateComment(
    commentId: string, 
    content: string, 
    type: "OFFICIAL" | "USER",
    is_adopted_user?: boolean
  ): Observable<any> {
    return this.http.put<any>(`${environment.apiUrl}/comments/${commentId}`, {
      content,
      type,
      is_adopted_user
    });
  }

  // Delete a comment
  deleteComment(commentId: string): Observable<any> {
    return this.http.delete<any>(`${environment.apiUrl}/comments/${commentId}`);
  }
}
